### PR TITLE
All Products block: Try hydrating

### PR DESCRIPTION
**Pull request by ockham**
WIP. Some early experiments, applying techniques from https://github.com/WordPress/block-hydration-experiments to the All Products block. See #20 for a vague roadmap.

To test, insert the "All Products" block into a page of your choice. View it on the frontend, and verify that it renders correctly and reacts to user interaction.

TODO:
- [x] Attributes aren't being serialized (as `data-gutenberg-attributes`). Find out why, fix.
- [x] Add frontend logic

### CODE REVIEW

**Comment by luisherranz** [comment]
Hey @ockham, would you mind updating the code (nomenclature) once we merge [this PR](https://github.com/WordPress/block-hydration-experiments/pull/45)? I don't want the `gunteberg-xxx` naming to spread if we are not finally going to use that 🙂

Thanks!!

**Comment by ockham** [comment]
FWIW, I spun off a PR against the upstream repo to carry over the migration of the All Products block to `block.json`: https://github.com/woocommerce/woocommerce-blocks/pull/6754

**Comment by ockham** [comment]
> Hey @ockham, would you mind updating the code (nomenclature) once we merge [this PR](https://github.com/WordPress/block-hydration-experiments/pull/45)? I don't want the `gunteberg-xxx` naming to spread if we are not finally going to use that 🙂
> 
> Thanks!!

Ooh, looks like I'd missed that! 😅 

Sure, I'll update that once it's in!

**Comment by ockham** [comment]
> > Hey @ockham, would you mind updating the code (nomenclature) once we merge [this PR](https://github.com/WordPress/block-hydration-experiments/pull/45)? I don't want the `gunteberg-xxx` naming to spread if we are not finally going to use that 🙂
> > Thanks!!
> 
> Ooh, looks like I'd missed that! 😅
> 
> Sure, I'll update that once it's in!

Done in 6ade0cf.

**Comment by ockham** [comment]
@michalczaplinski I've merged your branch from #39 into this one and applied some minor changes (23b0f51) to use 'our' hydration technique (from the Block Hydration Experiments repo). I _think_ it's basically working. Hooray? 🎉 😄 

(I should say that it's not a very big feat since the it's not that different from the hydration technique that was previously used by the block.)

(Edit: Currently getting a block validation error in the editor -- TBH I didn't care much for backcompat and proper deprecations for the sake of this experiment 😬 )

cc/ @sunyatasattva @gigitux

**Comment by michalczaplinski** [comment]
> Currently getting a block validation error in the editor

I think this is because of the [changes to the `save.js`](https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/pull/21/files#diff-63c8840763e14662950356f92ede146e565e5bc57d99fb0f2ca6923109ee346dL12-L20). You're not serializing the attributes in the `save()` anymore which makes sense because it would be redundant but at at the same time the saved markup is now different hence the validation error. I don't know which is more important: not being redundant or not showing validation errors but we can ask Woo and deal with it later 🙂 

I see that some e2e tests are also failing but I think that they are failing for all PRs so this is probably unrelated. I ll try to pull the latest changes from upstream `woocommerce-blocks` repo to see if this was resolved.

Also, I have manually disabled (from the github UI) the [`generate-zip`github action](https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/blob/f11f11c9ad91c611340c1ad64b7266aca391a72b/.github/workflows/generate-zip.yml#L1-L64) because it seems to depend on some secrets that we don't have in this repo.

**Comment by michalczaplinski** [comment]
In any case, it does indeed seem to work just fine 👏 😄 Nice work @ockham 

Are you planning to add more changes before we can merge this?

**Comment by ockham** [comment]
> > Currently getting a block validation error in the editor
> 
> I think this is because of the [changes to the `save.js`](https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/pull/21/files#diff-63c8840763e14662950356f92ede146e565e5bc57d99fb0f2ca6923109ee346dL12-L20). You're not serializing the attributes in the `save()` anymore which makes sense because it would be redundant but at at the same time the saved markup is now different hence the validation error. I don't know which is more important: not being redundant or not showing validation errors but we can ask Woo and deal with it later 🙂

Oh yeah, that's certainly the reason -- what I meant to say is that I haven't bothered to write a block deprecation to handle this just yet. I might do that though so we can merge the PR without sacrificing backwards compatibility -- should be fairly straight-forward 😄 

> I see that some e2e tests are also failing but I think that they are failing for all PRs so this is probably unrelated. I ll try to pull the latest changes from upstream `woocommerce-blocks` repo to see if this was resolved.

Thank you! AFAICS, they've been failing consistently in the upstream repo for a while 😬 (They sure were when I filed my [`block.json` refactor PR](https://github.com/woocommerce/woocommerce-blocks/pull/6754) last week.) 

> Also, I have manually disabled (from the github UI) the [`generate-zip`github action](https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/blob/f11f11c9ad91c611340c1ad64b7266aca391a72b/.github/workflows/generate-zip.yml#L1-L64) because it seems to depend on some secrets that we don't have in this repo.

👍

**Comment by ockham** [comment]
> In any case, it does indeed seem to work just fine 👏 😄 Nice work @ockham

Thank you! 😊 

> Are you planning to add more changes before we can merge this?

I'll add that block deprecation I mentioned in my other comment 😄

**Comment by ockham** [comment]
I now get

```
Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined.
```

every now and then on the frontend. Looks like some race condition to me, probably related to `registerBlockType` and the `StoreNoticesProvider` wrapper... 😕

**Comment by michalczaplinski** [comment]
> ```Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined.```

I have seen it one time as well when I ran your PR locally for the first time. But I restarted webpack and `wp-env` and it went away so I shrugged it off. Sounds like a race condition indeed. 

Could it be that the [`Comp` is not yet set in the `blockTypes` registry](https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/blob/fac49f7c5b7351d6e0f37b524115be606895be1c/assets/js/base/utils/bhe-frontend.tsx#L125-L126) by the time that we run the hydration code?

**Comment by michalczaplinski** [comment]
I can confirm that there is indeed a race condition:

https://user-images.githubusercontent.com/5417266/182431565-ad6d6aa5-4fb6-47fc-b410-7a4bf700a740.mp4

Whenever the `all-products-frontend` file is served from the memory cache of the browser there is no error. But when it's served from the network, there is an error. 

Also, you can just put a `console.log(Comp)` after this line https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/blob/fac49f7c5b7351d6e0f37b524115be606895be1c/assets/js/base/utils/bhe-frontend.tsx#L125 and you'll see that whenever the error appears it will log `undefined` but if there's no error it will log the `Comp`

So, in conclusion, we have to ensure that the block registration is always run before the hydration 🙂.

**Comment by ockham** [comment]
> I can confirm that there is indeed a race condition:
> 
>  2022-08-02_11-50-56.mp4 
> Whenever the `all-products-frontend` file is served from the memory cache of the browser there is no error. But when it's served from the network, there is an error.

Thanks a lot for confirming!

> Also, you can just put a `console.log(Comp)` after this line
> 
> https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/blob/fac49f7c5b7351d6e0f37b524115be606895be1c/assets/js/base/utils/bhe-frontend.tsx#L125
> 
> and you'll see that whenever the error appears it will log `undefined` but if there's no error it will log the `Comp`
> So, in conclusion, we have to ensure that the block registration is always run before the hydration 🙂.

Fun fact, I had a `console.log(Comp)` on that line, but on my Intel-based Mac, it always logged the component (and no error was triggered). I _think_ that `console.log()` slowed down execution enough that the component ended up being registered just in time 😅 

Now on my new M1 Pro, I get `undefined` sometimes 😌

**Comment by michalczaplinski** [comment]
So, I see two ways of handling this:

1. We could solve it ourselves by making the `registerBlock` asynchronous and then define a function like `whenRegistered` that resolves a promise when a particular block gets registered. This might look something like:

```diff
diff --git a/assets/js/base/utils/bhe-frontend.tsx b/assets/js/base/utils/bhe-frontend.tsx
index 318f02ce4..ba122be7f 100644
--- a/assets/js/base/utils/bhe-frontend.tsx
+++ b/assets/js/base/utils/bhe-frontend.tsx
@@ -72,7 +72,10 @@ Children.shouldComponentUpdate = () => false;
 
 class WpBlock extends HTMLElement {
 	connectedCallback() {
-		setTimeout( () => {
+		setTimeout( async () => {
+			const blockType = this.getAttribute( 'data-wp-block-type' );
+			await whenRegistered( blockType );
+
 			// ping the parent for the context
 			const event = new CustomEvent( 'wp-block-context', {
 				detail: {},
```

The caveat is that we might have to add a timeout of sorts in case the block fails to register so that this `await` does not get stuck forever. But that should not be a big issue (I guess) because if the block never gets registered for some reason (script fails to load, etc.) then we cannot hydrate anyway.

2. We could make sure that the `frontend.js` which contains the `registerBlock(MyBlock)` is enqueued before the frontend file that contains the `wp-block` and the hydration code. I don't know enough about those enqueing mechanisms to know if this is feasible or a good idea.

What do you think @ockham ? 🙂

**Comment by ockham** [comment]
Thanks a lot for your investigation @michalczaplinski -- much appreciated! ❤️ 

When I was thinking about the issue earlier, I was also wondering if manually introducing some asynchronicity could solve the problem. TBH I didn’t quite see how we can create a promise during block registration that the custom element listens to 🙈  but I’m super curious to see what solution you come up with! 😄 

I think I was personally leaning more towards solution #2. IIRC, the block’s `frontend.js` is run differently (via an existing Woo-specific mechanism) than we do in BHE, so it’s no surprise if it’s not quite compatible with our approach.

**Comment by ockham** [comment]
> I think I was personally leaning more towards solution #2. IIRC, the block’s `frontend.js` is run differently (via an existing Woo-specific mechanism) than we do in BHE, so it’s no surprise if it’s not quite compatible with our approach.

I looked it up: So in BHE, we're using the `block.json`'s `viewScript` field to load `frontend.js` ([e.g.](https://github.com/WordPress/block-hydration-experiments/blob/8ebdddf255a155b365a0be75df39757900ad4930/src/blocks/interactive-parent/block.json#L45)).

I haven't totally tracked it down in Woo Blocks, but I've found at least some related [Webpack logic](https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/blob/f11f11c9ad91c611340c1ad64b7266aca391a72b/bin/webpack-entries.js#L7-L11) and [documentation](https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/blob/f11f11c9ad91c611340c1ad64b7266aca391a72b/docs/contributors/contributing/block-assets.md).

It seems like those frontend scripts get enqueued via [`AbstractBlocks`](https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/blob/f11f11c9ad91c611340c1ad64b7266aca391a72b/src/BlockTypes/AbstractBlock.php) (and its derived classes). Maybe this happens a bit "late"? Might be worth a try using `viewScript` instead 🤔

**Comment by michalczaplinski** [comment]
Thanks for digging a bit deeper @ockham ! 

We took a look with Carlos and this is what we found:

1. We tried to use the to use the block.json's `viewScript` field but to no avail. I think this might be because Woocommerce overrides the default webpack configuration so it doesn't look for those files defined in the `viewScript` anymore.

2. We pinned down the problem a little more precisely. The issue is that the `registerBlockType()` ends up in the `all-products-frontend` bundle but for some reason it's placed in the bundle **after** the `customElements.define('wp-block', WpBlock);`:

   https://user-images.githubusercontent.com/5417266/182756559-3357122f-a74c-4474-ab78-43e0e7f233bf.mp4

----

I can't couldn't think of another way to solve this problem today so I went ahead and wrapped the block registry in a class that implements a `whenRegistered()` method (akin to the [`whenDefined()` method](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/whenDefined) of the `customElementRegistry`. The diff is in https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/pull/21/commits/8edc7b2f2115c21fb11dcee86e92d24017b4c305.

It solves the bug and I think the code is not all that terrible 😅  I can also put it a separate PR!

**Comment by ockham** [comment]
> Thanks for digging a bit deeper @ockham !
> 
> We took a look with Carlos and this is what we found:
> 
> 1. We tried to use the to use the block.json's `viewScript` field but to no avail. I think this might be because Woocommerce overrides the default webpack configuration so it doesn't look for those files defined in the `viewScript` anymore.

Indeed: Woo seems to have an allowlist of scripts to look out for in a given block directory, and only those files will be built by Webpack. (Here's the code for [`frontend.js`](https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/blob/f11f11c9ad91c611340c1ad64b7266aca391a72b/bin/webpack-entries.js#L129-L134), for example.)

> 2. We pinned down the problem a little more precisely. The issue is that the `registerBlockType()` ends up in the `all-products-frontend` bundle but for some reason it's placed in the bundle **after** the `customElements.define('wp-block', WpBlock);`:

Great sleuthing! 👍 🕵️ 

> I can't couldn't think of another way to solve this problem today so I went ahead and wrapped the block registry in a class that implements a `whenRegistered()` method (akin to the [`whenDefined()` method](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/whenDefined) of the `customElementRegistry`. The diff is in [8edc7b2](https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/commit/8edc7b2f2115c21fb11dcee86e92d24017b4c305).
> 
> It solves the bug and I think the code is not all that terrible 😅 I can also put it a separate PR!

Impressive, thanks for that solution!

**Comment by ockham** [comment]
Being somewhat of a die-hard minimalist (occam's razor and all 😉), I had another look at the problem, based on your findings above 😬 

Essentially, `customElements.define('wp-block', WpBlock);` isn't wrapped in any function or class, so it's run as a side-effect whenever `bhe-frontend.tsx` is imported by another file. This seems problematic, since our `registerBlockType()` definition also lives in that file.

However, the same is true of the corresponding file in BHE, so apparently, simply importing `registerBlockType()` is not a problem (it's going to be used by the file that imports it to register the relevant block type, which I guess happens in time before the `customElements.define()` call.)

What would seem more problematic is if `bhe-frontend.tsx` was imported by a file that doesn't call `registerBlockType()` right away, since that would certainly run the `customElements.define()` side effect _before_ the block is registered.

Well, turns out we're kinda doing that: https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/pull/21/files#diff-b1f2eacd00a094cfdf9ca09fbfb8ea734306805ada4af1ef6042bec87c617405R10 😬 

That `utils` file is imported by a lot of blocks. Does that mean we end up with a ton of `customElements.define()` calls all over the place in our build?

image

Looks like 😬 

For this PR, we're only importing `registerBlockType` directly from `bhe-frontend` though, not via `util` -- so that line in util isn't really needed. What happens if we remove it?

```diff
diff --git a/assets/js/base/utils/index.js b/assets/js/base/utils/index.js
index 5f1aa8fd1..802954efe 100644
--- a/assets/js/base/utils/index.js
+++ b/assets/js/base/utils/index.js
@@ -7,5 +7,5 @@ export * from './get-valid-block-attributes';
 export * from './product-data';
 export * from './derive-selected-shipping-rates';
 export * from './get-icons-from-payment-methods';
-export * from './bhe-frontend';
+//export * from './bhe-frontend';
 export * from './bhe-element';
```

image

Better 😄 

---

If this solution seems fragile (loading order and all), I'm pretty sure we can make it solid: We can isolate the `customElements.define()` side effect in a file that's never directly imported by the block developer, but only used by "the framework" (eventually Gutenberg, but for now WooCommerce Blocks) upon loading the frontend (think some `wp_enqueue_script`). In any case, I think we can do without a dedicated block registry 😅

**Comment by michalczaplinski** [comment]
Awesome work @ockham ! 👏 

I had a lingering suspicion that this should be solved by being careful about importing modules but I didn't see how to approach it, major kudos! 🙂 

With that out of the way, I think we can merge it now 🙂

**Review by michalczaplinski — APPROVED** [review]

**Comment by luisherranz** [comment]
Sorry for being a bit late to the party, but this race condition was expected, and it is what I meant with the _Make sure Frontend components are automatically hydrated even if their component is registered after the connectedCallback execution_ task in the [BHE Tracking issue](https://github.com/WordPress/block-hydration-experiments/issues/39).

I'll open an issue in the BHE repository.

**Comment by luisherranz** [comment]
Done:
- https://github.com/WordPress/block-hydration-experiments/issues/52
